interfaces/apparmor/apparmor.go

```diff
@@ -37,46 +37,69 @@ import (
 	"github.com/snapcore/snapd/osutil"
 )
 
-// LoadProfile loads an apparmor profile from the given file.
+// LoadProfile loads apparmor profiles from the given files.
 //
 // If no such profile was previously loaded then it is simply added to the kernel.
 // If there was a profile with the same name before, that profile is replaced.
-func LoadProfile(fname string) error {
-	return loadProfile(fname, dirs.AppArmorCacheDir)
+func LoadProfiles(fnames []string) error {
+	return loadProfiles(fnames, dirs.AppArmorCacheDir)
 }
 
 // UnloadProfile removes the named profile from the running kernel.
 //
 // The operation is done with: apparmor_parser --remove $name
 // The binary cache file is removed from /var/cache/apparmor
-func UnloadProfile(name string) error {
-	return unloadProfile(name, dirs.AppArmorCacheDir)
+func UnloadProfiles(names []string) error {
+	return unloadProfiles(names, dirs.AppArmorCacheDir)
 }
 
-func loadProfile(fname, cacheDir string) error {
+func loadProfiles(fnames []string, cacheDir string) error {
+	if len(fnames) == 0 {
+		return nil
+	}
+
 	// Use no-expr-simplify since expr-simplify is actually slower on armhf (LP: #1383858)
-	args := []string{"--replace", "--write-cache", "-O", "no-expr-simplify", fmt.Sprintf("--cache-loc=%s", cacheDir)}
+	args := []string{"--replace", "--write-cache", "-O", "no-expr-simplify",
+		fmt.Sprintf("--cache-loc=%s", cacheDir)}
+
 	if !osutil.GetenvBool("SNAPD_DEBUG") {
 		args = append(args, "--quiet")
 	}
-	args = append(args, fname)
+	args = append(args, fnames...)
 
 	output, err := exec.Command("apparmor_parser", args...).CombinedOutput()
 	if err != nil {
-		return fmt.Errorf("cannot load apparmor profile: %s\napparmor_parser output:\n%s", err, string(output))
+		if len(fnames) > 1 {
+			// Revert possibly applied profiles. Note that this will always return
+			// an error as at least one of the profiles could not be loaded. We do
+			// not log it as this is anyway a paliative solution.
+			args := []string{"--remove"}
+			args = append(args, fnames...)
+			exec.Command("apparmor_parser", args...).Run()
+		}
+		return fmt.Errorf("cannot load apparmor profile: %s\napparmor_parser output:\n%s",
+			err, string(output))
 	}
 	return nil
 }
 
-func unloadProfile(name, cacheDir string) error {
-	output, err := exec.Command("apparmor_parser", "--remove", name).CombinedOutput()
+func unloadProfiles(names []string, cacheDir string) error {
+	if len(names) == 0 {
+		return nil
+	}
+
+	args := []string{"--remove"}
+	args = append(args, names...)
+	output, err := exec.Command("apparmor_parser", args...).CombinedOutput()
```

Comment:
When profiles are removed and an error occurs, does the removal stop on the first error and any following profiles are still loaded or does it go through all of them anyway?

Reply:
apparmor_parser removes/adds all profiles it can, even if some of them fail.

Reply:
FYI, parallel profile loads was introduced in upstream apparmor 2.11 (https://gitlab.com/apparmor/apparmor/wikis/Release_Notes_2.11) in https://gitlab.com/apparmor/apparmor/commit/3cb1477f5d046fd957f26145a9a4eb444ef57f43. AppArmor 2.10.95 (Ubuntu 14.04 and 16.04) is a pre-release version which contains this functionality. Debian Stretch (the earliest version with snapd) contains 2.11. I believe it was this release that made the loading/removing robust and fixed the issue where apparmor_parser would short-circuit on profile error (eg, while you can give apparmor 2.7 a list of file, if one fails to compile it exits with error and all other profiles are left unloaded).

I've verified that Ubuntu 14.04 - 18.10 correctly loads/removes all the profiles it can and does not short-circuit on error.

In terms of supported Ubuntu and Debian releases, this looks ok, but someone would need to verify other distributions that load apparmor profiles.